### PR TITLE
[224] l build non uk degree details page

### PR DESCRIPTION
**Pull request by JonF1**
### Context
* https://trello.com/c/Iwg4wgLH/224-l-build-non-uk-degree-details-page

Adding the non-UK degree details page to education history user journey. 

### Changes proposed in this pull request
Renders uk or non-uk degree details partial dependent on degree type
Saves degree with non-UK params in db
Adds non-UK degree type validations
Country data for country select box 
Feature, model tests to verify correct behaviour 
Refactor views with helpers  
Non-UK trait in degree factory
### Guidance to review
Create a degree from the overview page and follow the journey. Continue button takes you back to overview as confirmation page not in master/in progress
https://dfe-twd-rttd-pr-103.herokuapp.com/

image

image

### CODE REVIEW

File: app/helpers/degrees_helper.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+private
+
+  def boilerplate(array)
+    result = array.map do |name|
+      OpenStruct.new(name: name)
+    end
+    result.unshift(OpenStruct.new(name: nil))
+    result
+  end
+end
```
Comment (aliuk2012) [inline app/helpers/degrees_helper.rb:30]: We don't seem to have any tests for this file

File: app/views/trainees/degrees/_uk_degree_form.html.erb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      <%= f.govuk_radio_button :degree_grade, name, label: { text: name }, link_errors: true %>
+    <% end %>
+    <%= f.govuk_radio_divider %>
+    <%= f.govuk_radio_button :degree_grade, :still_studying, 
+          label: { text: "The trainee is still studying for their degree"} %>
+  <% end %>
+
+  <%= f.govuk_submit("Continue") %>
+<% end %>
+
```
Comment (aliuk2012) [inline app/views/trainees/degrees/_uk_degree_form.html.erb:29]: 🙈

File: app/views/trainees/degrees/_uk_degree_form.html.erb [hunk 3]
```diff
@@ -0,0 +1,29 @@
+<%= form_with model: [ @trainee, @degree ], local: true do |f| %>
+  <h2 class="govuk-heading-l">Degree Details</h2>
```
Comment (aliuk2012) [inline app/views/trainees/degrees/_uk_degree_form.html.erb:2]:
```suggestion
  <h1 class="govuk-heading-l">Degree Details</h1>
```

File: db/schema.rb [hunk 4]
```diff
@@ -75,15 +74,14 @@
...
-    t.text "middle_names"
     t.integer "record_type"
+    t.text "middle_names"
     t.text "international_address"
     t.integer "locale_code"
     t.integer "gender"
     t.index ["dttp_id"], name: "index_trainees_on_dttp_id"
     t.index ["gender"], name: "index_trainees_on_gender"
     t.index ["locale_code"], name: "index_trainees_on_locale_code"
-    t.index ["record_type"], name: "index_trainees_on_record_type"
```
Comment (aliuk2012) [inline db/schema.rb:86]: Why are we removing record_type index? Theres a couple other weird things going on here. I thought the migration was only going to add country to degrees.

Also did you run db:migrate? I was expecting to see the database diagram to be included in this PR

File: app/views/trainees/degrees/_uk_degree_form.html.erb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        label: { text: "Graduation Year", size: 's' }, 
+        width: "one-quarter"  %>
+
+  <%= f.govuk_radio_buttons_fieldset(:degree_grade, 
+        legend: { text: "Degree grade", size: 's' }, 
+        classes: "degree-grade" ) do %>
+
+    <% DEGREE_GRADES.map do |name| %>
+      <%= f.govuk_radio_button :degree_grade, name, label: { text: name }, link_errors: true %>
+    <% end %>
```
Comment (aliuk2012) [inline app/views/trainees/degrees/_uk_degree_form.html.erb:21]: should this not be part of the helper method?

File: app/views/trainees/degrees/_non_uk_degree_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,16 @@
+<%= form_with model: [ @trainee, @degree ], local: true do |f| %>
+  <%= f.govuk_error_summary %>
+  <h2 class="govuk-heading-l">Degree Details</h2>
+    <%= f.govuk_collection_select :degree_subject, degree_subjects,
```
Comment (darokel) [inline app/views/trainees/degrees/_non_uk_degree_form.html.erb:4]: slighty off indentation here

File: app/views/trainees/degrees/_uk_degree_form.html.erb (hunk 2 above)
Comment (JonF1) [reply app/views/trainees/degrees/_uk_degree_form.html.erb]: fixed

File: app/views/trainees/degrees/_uk_degree_form.html.erb (hunk 3 above)
Comment (JonF1) [reply app/views/trainees/degrees/_uk_degree_form.html.erb]: fixed

File: app/views/trainees/degrees/_non_uk_degree_form.html.erb (hunk 6 above)
Comment (JonF1) [reply app/views/trainees/degrees/_non_uk_degree_form.html.erb]: fixed

File: app/helpers/degrees_helper.rb (hunk 1 above)
Comment (defong) [reply app/helpers/degrees_helper.rb]: Done
- to the extent of circumventing the statics lists (HESA_DEGREE_TYPES, INSTITUTIONS, DEGREE_SUBJECTS, DEGREE_COUNTRIES) to prove a plus one item is returned

File: app/views/trainees/degrees/_uk_degree_form.html.erb (hunk 5 above)
Comment (defong) [reply app/views/trainees/degrees/_uk_degree_form.html.erb:21]: need to pick your brain for that one....

**Review by darokel — APPROVED** [review]

File: app/views/trainees/degrees/_non_uk_degree_form.html.erb [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+			hint: { text: 'Search for the closest matching subject' } %>
+
+	<%= f.govuk_collection_select :country, degree_countries_options, :name,
+			:name, label: { text: "Country where the degree was obtained", size: 's' } %>
+
+	<%= f.govuk_text_field :graduation_year, 
+			label: { text: "Graduation Year", size: 's' }, 
+			width: "one-quarter"  %>
+
+	<%= f.govuk_submit("Continue") %>
```
Comment (aliuk2012) [inline app/views/trainees/degrees/_non_uk_degree_form.html.erb:15]: indenting issues

File: app/views/trainees/degrees/_uk_degree_form.html.erb (hunk 5 above)
Comment (aliuk2012) [reply app/views/trainees/degrees/_uk_degree_form.html.erb:21]: I'm assuming these will most likely end up being saved in the db so I'm happy for us to merge this as is

File: app/views/trainees/degrees/_uk_degree_form.html.erb [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    <%= f.govuk_radio_divider %>
+    <%= f.govuk_radio_button :degree_grade, :still_studying, 
+          label: { text: "The trainee is still studying for their degree"} %>
+  <% end %>
+
+  <%= f.govuk_submit("Continue") %>
+<% end %>
```
Comment (aliuk2012) [inline app/views/trainees/degrees/_uk_degree_form.html.erb:28]: 🙈

Comment (JonF1) [reply app/views/trainees/degrees/_uk_degree_form.html.erb:27]: Was it the extra line here? I've removed

File: app/views/trainees/degrees/_non_uk_degree_form.html.erb (hunk 7 above)
Comment (JonF1) [reply app/views/trainees/degrees/_non_uk_degree_form.html.erb]: fixed, cheers

**Review by aliuk2012 — APPROVED** [review]

**Pull request merged**